File: JavaScript/Tree/BST.js
class Node {
    constructor(data) {
        this.data = data
        this.right = null
        this.left = null
    }
}

class BST {
    constructor() {
        this.root = null
    }

    insertionHelper(root, data) {
        if (root.data < data) {
            if (root.right === null) {
                root.right = new Node(data)
                return root.right
            }
            else {
                return this.getLeaf(root.right, data)
            }
        }
        else {
            if (root.left === null) {
                root.left = new Node(data)
                return root.left
            }
            else {
                return this.getLeaf(root.left, data)
            }
        }
    }

    insert(data) {
        if (!this.root) {
            this.root = Node(data)
            return
        }

        this.insertionHelper(this.root, data)
    }

    successor(root) {
        let node = root.left

        while (node != null) {
            node = node.left
        }

        return node.data
    }

    predecessor(root) {
        let node = root.right

        while (node != null) {
            node = node.right
        }

        return node.data
    }

    deleteHelper(root, data) {
        if (root === null) return root
        else if (root.data < data) root.right = this.deleteHelper(root.right, data)
        else if (root.data > data) root.left = this.deleteHelper(root.left, data)
        else {
            if (root.right === null && root.left === null) {
                root = null
            }
            else if (root.right !== null) {
                root.data = this.successor(root)
                root.right = this.deleteHelper(root.right, root.data)
            }
            else {
                root.data = this.predecessor(root)
                root.left = this.deleteHelper(root.left, root.data)
            }
        }

        return root
    }

    delete(data) {
        let root = search(data)

        return this.deleteHelper(root, data)
    }

    search(data) {
        if (root.data === data) return root
        else if (root.data > data) return this.search(root.left, data)
        else if (root.data < data) return this.search(root.right, data)
    }
}

File: CPP/DS/BST.cpp
#include <iostream>

template <typename T>
class Node
{
public:
    T data;
    Node<T> *left;
    Node<T> *right;

    Node(T element)
    {
        data = element;
        left = nullptr;
        right = nullptr;
    }
};

template <typename T>
class BST
{
private:
    Node<T> *root;

    void pre_order(Node<T> *root) const
    {
        if (!root)
            return;
        std::cout << root->data << " ";
        in_order_helper(root->left);
        in_order_helper(root->right);
    }

    void in_order(Node<T> *root) const
    {
        if (!root)
            return;
        in_order_helper(root->left);
        std::cout << root->data << " ";
        in_order_helper(root->right);
    }

    void post_order(Node<T> *root) const
    {
        if (!root)
            return;
        in_order_helper(root->left);
        in_order_helper(root->right);
        std::cout << root->data << " ";
    }

    T successor(Node<T> *root)
    {
        Node<T> *node = root->right;
        while (node && node->left)
        {
            node = node->left;
        }
        return node->data;
    }

    T predecessor(Node<T> *root)
    {
        Node<T> *node = root->left;
        while (node && node->right)
        {
            node = node->right;
        }
        return node->data;
    }

    Node<T> *insert_helper(Node<T> *root, T data)
    {
        if (!root)
        {
            return new Node<T>(data);
        }

        if (data > root->data)
        {
            root->right = insert_helper(root->right, data);
        }
        else if (data < root->data)
        {
            root->left = insert_helper(root->left, data);
        }
        else
        {
            return root;
        }
    }

    Node<T> *remove_helper(Node<T> *root, T data)
    {
        if (!root)
        {
            return root;
        }
        else if (data > root->data)
        {
            root->right = remove_helper(root->right, data);
        }
        else if (data < root->data)
        {
            root->left = remove_helper(root->left, data);
        }
        else
        {
            //* Node is leaf
            if (!root->right && !root->left)
            {
                delete root;
                return nullptr;
            }
            //* Node doesnt have left.
            else if (!root->left)
            {
                Node<T> *temp = root->right;
                delete root;
                return temp;
            }
            //* Node doesnt have right.
            else if (!root->right)
            {
                Node<T> *temp = root->left;
                delete root;
                return temp;
            }
            //* Node have both right and left.
            else
            {
                root->data = successor(root);
                root->right = remove_helper(root->right, root->data);
            }
        }

        return root;
    }

    void destroy_helper(Node<T> *root)
    {
        if (!root)
            return;

        destroy_helper(root->left);
        destroy_helper(root->right);

        delete root;
    }

public:
    template <typename U>
    friend std::ostream &operator<<(std::ostream &os, const AVL<U> &tree);

    BST()
    {
        root = nullptr;
    }

    BST(std::initializer_list<T> init_list)
    {
        root = nullptr;
        for (T element : init_list)
        {
            insert(element);
        }
    }

    void insert(T element)
    {
        root = insert_helper(root, element);
    }

    void remove(T element)
    {
        root = remove_helper(root, element);
    }

    void print() const
    {
        in_order(root);
        std::cout << std::endl;
    }

    ~AVL()
    {
        destroy_helper(root);
    }
};

template <typename T>
std::ostream &operator<<(std::ostream &os, const BST<T> &tree)
{
    tree.print();
    return os;
}